File: mvpboilerplate/src/main/java/com/agit/lutluthfi/mvpboilerplate/utils/rx/SchedulerProvider.java
package com.agit.lutluthfi.mvpboilerplate.utils.rx;

import io.reactivex.CompletableTransformer;
import io.reactivex.FlowableTransformer;
import io.reactivex.MaybeTransformer;
import io.reactivex.ObservableTransformer;
import io.reactivex.SingleTransformer;

public interface SchedulerProvider {

    <T> ObservableTransformer<T, T> ioToMainObservableScheduler();

    <T> FlowableTransformer<T, T> ioToMainFlowableScheduler();

    <T> SingleTransformer<T, T> ioToMainSingleScheduler();

    <T> MaybeTransformer<T, T> ioToMainMaybeScheduler();

    CompletableTransformer ioToMainCompletableScheduler();
}

File: mvpboilerplate/src/main/java/com/agit/lutluthfi/mvpboilerplate/SchedulerProvider.kt
package com.agit.lutluthfi.mvpboilerplate

import io.reactivex.CompletableTransformer
import io.reactivex.FlowableTransformer
import io.reactivex.MaybeTransformer
import io.reactivex.ObservableTransformer
import io.reactivex.SingleTransformer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers

object SchedulerProvider {

    fun <T> ioToMainObservableScheduler(): ObservableTransformer<T, T> =
            ObservableTransformer { upstream ->
                upstream.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
            }

    fun <T> ioToMainSingleScheduler(): SingleTransformer<T, T> =
            SingleTransformer { upstream ->
                upstream.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
            }


    fun ioToMainCompletableScheduler(): CompletableTransformer =
            CompletableTransformer { upstream ->
                upstream.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
            }


    fun <T> ioToMainFlowableScheduler(): FlowableTransformer<T, T> =
            FlowableTransformer { upstream ->
                upstream.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
            }


    fun <T> ioToMainMaybeScheduler(): MaybeTransformer<T, T> =
            MaybeTransformer { upstream ->
                upstream.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
            }
}
